### PR TITLE
fix: support empty state document

### PR DESCRIPTION
**Pull request by jcosentino11**
**Issue #, if available:**

**Description of changes:**
Support an empty shadow document,`{"state": {}}`.

**Why is this change necessary:**
According to https://docs.aws.amazon.com/iot/latest/developerguide/device-shadow-document.html#device-shadow-empty-fields, `{}` is a valid shadow document.

**How was this change tested:**
New unit and integration tests

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [x] Updated or added new unit tests
 - [x] Updated or added new integration tests
 - [ ] If your code makes a remote network call, it was tested with a proxy
 
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by MikeDombo — DISMISSED** [review]

File: src/test/java/com/aws/greengrass/shadowmanager/util/JsonUtilTest.java [hunk 1]
```diff
@@ -42,7 +42,6 @@ void setup() throws IOException {
     @ValueSource(strings={
             "{\"version\": 1}",
             "{}",
-            "{\"state\": {}}",
```
[the remark this answers is not in the document]
Comment (jcosentino11) [reply src/test/java/com/aws/greengrass/shadowmanager/util/JsonUtilTest.java:45]: According to docs I don't believe so, but I'll make sure to confirm with manual testing

**Review by saranyailla — APPROVED** [review]

File: src/test/java/com/aws/greengrass/shadowmanager/util/JsonUtilTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jcosentino11) [reply src/test/java/com/aws/greengrass/shadowmanager/util/JsonUtilTest.java:45]: testing in IoT console and cli, updating a shadow with`null` is valid, but subsequent gets will return `{}`

[the remark this answers is not in the document]
Comment (MikeDombo) [reply src/test/java/com/aws/greengrass/shadowmanager/util/JsonUtilTest.java:45]: OK, so we need to support that too then.

And how about just `{}` without `{"state": <something>}`

[the remark this answers is not in the document]
Comment (jcosentino11) [reply src/test/java/com/aws/greengrass/shadowmanager/util/JsonUtilTest.java:45]: yep just added `null` support.  

{} without state is invalid:
```The Device Shadow couldn't be updated. Missing required node: state```

**Review by MikeDombo — APPROVED** [review]

File: src/main/resources/json/schema/update_payload_schema.json [hunk 2]
```diff
@@ -11,7 +11,7 @@
     },
     "state": {
       "description": "The state node containing the desired, reported, and delta shadow state.",
-      "type": "object",
+      "type": ["object", "null"],
```
Comment (MikeDombo) [inline src/main/resources/json/schema/update_payload_schema.json:14]: so we're considering these values as valid now, which is fine, but are we actually handling the values correctly?

What is our behavior after validating the data?

**Comment by jcosentino11** [comment]
Adding more tests for local to cloud sync, realized that merging logic needs to be adjusted too

File: src/main/resources/json/schema/update_payload_schema.json (hunk 2 above)
Comment (jcosentino11) [reply src/main/resources/json/schema/update_payload_schema.json:14]: yeah realized this after more testing, currently reviewing/updating merging logic to handle null/empty state

File: src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java [hunk 3]
```diff
@@ -58,6 +58,11 @@ public ShadowState deepCopy() {
      */
     @SuppressWarnings({"PMD.ForLoopCanBeForeach", "PMD.NullAssignment"})
     public void update(JsonNode updatedStateNode) {
+        if (isNullOrMissing(updatedStateNode)) {
+            this.desired = null;
+            this.reported = null;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java:63]: why are both being set to null?

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java:63]: My understanding is that if shadow is being updated (either cloud or local) with "state: {}" or "state: null", then that signals we want to "clear" the document, so setting desired/reported to null

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java:63]: Need to verify that. Updating the shadow state should only impact the reported state if the update is coming from the device, IMO.

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java:63]: To clear just the reported or just the desired state, it'd be `{"state": {"reported": null}`, so sending a `{"state": {}}` seems pretty intentional for clearing the full shadow, although destructive.  I don't think we can truly disambiguate if devices intended to clear reported or desired, since customers can modify both reported or desired state (even if they shouldn't).

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java:63]: OK, I see.

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java:63]: And do we handle that case?

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/shadowmanager/model/ShadowState.java:63]: Yeah null reported/desired is already expected/handled throughout the code

**Review by MikeDombo — APPROVED** [review]

**Review by saranyailla — APPROVED** [review]
